worker-workflow/src/main/java/com/github/cafdataprocessing/workflow/WorkflowSettingsRetriever.java

```diff
@@ -0,0 +1,264 @@
+                    throw (ApiException) ex.getCause();
+                } else if (ex.getCause() instanceof DocumentWorkerTransientException) {
+                    throw (DocumentWorkerTransientException) ex.getCause();
+                } else {
+                    throw new RuntimeException(ex);
+                }
+            }
+        }
+        return customConfigs;
+    }
+
+    private Map<String, String> processRepositoryConfigs(final Document document, final String tenantId,
+                                                         final Map<String, RepoConfigSource> configs) throws ApiException,
+                                                                                                             DocumentWorkerTransientException
+    {
+        final Map<String, String> customConfigs = new HashMap<>();
+        String repositoryKey = null;
+        for (final Map.Entry<String, RepoConfigSource> config : configs.entrySet()) {
+            switch (config.getValue().getSource()) {
+                case FIELD:
+                    repositoryKey = document.getField(
+                        config.getValue().getKey()).getValues().stream().findFirst().get().getStringValue();
+                    break;
+                case CUSTOMDATA:
+                    repositoryKey = document.getCustomData(config.getValue().getKey());
+                    if (repositoryKey == null) {
+                        throw new RuntimeException("Unable to obtain repository id for config " + config.getValue().getKey());
+                    }
+                    break;
+                default:
+                    throw new RuntimeException("Invalid source for repository id. Source of "
+                        + config.getValue().getSource() + " is not recognised as a valid source.");
+            }
+            try {
+                customConfigs.put(config.getKey(),
+                                  settingsCache.get(
+                                      new WorkflowSettingsRepositoryCacheKey(tenantId, config.getKey(), repositoryKey)));
+            } catch (final ExecutionException ex) {
+                if (ex.getCause() instanceof ApiException) {
+                    throw (ApiException) ex.getCause();
+                } else if (ex.getCause() instanceof DocumentWorkerTransientException) {
+                    throw (DocumentWorkerTransientException) ex.getCause();
+                } else {
+                    throw new RuntimeException(ex);
+                }
+            }
+        }
+        return customConfigs;
+    }
+
+    private static Map<String, String> processTaskConfigs(final Document document, final List<String> configs)
+        throws ApiException, DocumentWorkerTransientException
+    {
+        final Map<String, String> customConfigs = new HashMap<>();
+        for (final String config : configs) {
+            final String configValue = document.getCustomData("TASK_SETTING_" + config.toUpperCase(Locale.US));
```

Comment:
If the custom data is missing can we throw an exception with a specific message?

Reply:
We can't throw an exception here because it doesn't always have to be here, like is the case with sources. Most times the TaskSetting won't be present so the RepositorySetting will be used.